### PR TITLE
Domain estimation stratified sample

### PR DESCRIPTION
**Pull request by smishr**
I have implemented domain mean for `StratifiedSample` class. Variance is not done yet. Ideas were taken from and discussed in #71

### CODE REVIEW

File: src/SurveyDesign.jl [hunk 1]
```diff
@@ -150,6 +150,10 @@ struct StratifiedSample <: AbstractSurveyDesign
             data[!, :weights] = weights
             data[!, :probs] = 1 ./ data[!, :weights]
         end
+        data[!, :sampsize] = sampsize
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:153]: This might not be necessary. In most calculations we are using `sampfraction` and in some (such as calculating variance of the mean) we also need `popsize`. We can keep it for now and remember to check if we need it after we implement most functionality.

File: src/ht.jl [hunk 2]
```diff
@@ -30,14 +30,11 @@ function ht_svymean(x::Symbol, design)
     total_df = ht_svytotal(x, design)
     total = total_df.total[1]
     var_total = total_df.se[1]
-    
     mean = 1.0 ./ sum(design.popsize) .* total
+    # TODO check standard error, it is incorrect?
     # @show total_df, total, var_total, mean
     #1.0 ./ (sum(design.popsize).^2) .* 
-    
+    # @show design.popsize
```
Comment (iuliadmtru) [inline src/ht.jl:37]: Remove this.

File: src/ht.jl [hunk 3]
```diff
@@ -30,14 +30,11 @@ function ht_svymean(x::Symbol, design)
     total_df = ht_svytotal(x, design)
     total = total_df.total[1]
     var_total = total_df.se[1]
-    
     mean = 1.0 ./ sum(design.popsize) .* total
+    # TODO check standard error, it is incorrect?
```
Comment (iuliadmtru) [inline src/ht.jl:34]: Is it incorrect or not?

File: src/svyby.jl [hunk 4]
```diff
@@ -37,3 +37,14 @@ function svyby(formula::Symbol, by::Symbol, design::AbstractSurveyDesign, func::
     gdf = groupby(design.data, by)
     return combine(gdf, [formula, :weights] => ((a, b) -> func(a, design, b, params...)) => AsTable)
 end
+
+"""
+    svyby(formula, by, design, function, params)
```
Comment (iuliadmtru) [inline src/svyby.jl:42]: The signature needs to be updated as well, `params` doesn't exist anymore.

File: src/svyby.jl [hunk 5]
```diff
@@ -37,3 +37,14 @@ function svyby(formula::Symbol, by::Symbol, design::AbstractSurveyDesign, func::
...
+
+Generate subsets of a StratifiedSample.
+"""
+function svyby(formula::Symbol, by::Symbol, design::StratifiedSample, func::Function)
+    # TODO: add functionality for `formula::AbstractVector`
+    gdf_domain = groupby(design.data, by)
+    return combine(gdf_domain, [formula, :sampfraction, design.strata] => ((a,b,c) -> func(a,b,c,design)) => AsTable )
```
Comment (iuliadmtru) [inline src/svyby.jl:49]: `combine` here doesn't take `weights` as argument anymore, but we might need it for `svytotal`. We'll see when we implement `svyby` for `svytotal`.

File: src/svymean.jl [hunk 6]
```diff
@@ -68,9 +68,9 @@ function svymean(x::Vector{Symbol}, design::SimpleRandomSample)
 end
 
 """
-Inner method for `svyby`.
+Inner method for `svyby` to calculate standard error of domain mean of SimpleRandomSample.
```
Comment (iuliadmtru) [inline src/svymean.jl:71]: The extra description is not necessary here. We should describe the basic functionality above the main method and for other methods only mention what is special (if it's not obvious). In this case I would keep the old docstring.

File: src/svymean.jl [hunk 7]
```diff
@@ -84,15 +84,47 @@ function sem_svyby(x::AbstractVector, design::SimpleRandomSample, _)
 end
 
 """
-Inner method for `svyby`.
+Inner method for `svyby` to calculate standard error of domain mean of StratifiedSample.
```
Comment (iuliadmtru) [inline src/svymean.jl:87]: Here same as line 71.

File: src/svymean.jl [hunk 8]
```diff
@@ -84,15 +84,47 @@ function sem_svyby(x::AbstractVector, design::SimpleRandomSample, _)
...
 """
-function svymean(x::AbstractVector, design::SimpleRandomSample, weights)
-    return DataFrame(mean = mean(x), sem = sem_svyby(x, design::SimpleRandomSample, weights))
+function sem_svyby(x::AbstractVector, design::StratifiedSample,weights)
```
Comment (iuliadmtru) [inline src/svymean.jl:89]: What is the purpose of this function?

File: src/svymean.jl [hunk 9]
```diff
@@ -84,15 +84,47 @@ function sem_svyby(x::AbstractVector, design::SimpleRandomSample, _)
 end
 
 """
-Inner method for `svyby`.
+Inner method for `svyby` to calculate standard error of domain mean of StratifiedSample.
 """
-function svymean(x::AbstractVector, design::SimpleRandomSample, weights)
```
Comment (iuliadmtru) [inline src/svymean.jl:89]: It looks like you removed the `svymean` method for `svyby` for `SimpleRandomSample` and you didn't replace it with anything else. Why did you remove it? It was working fine for SRS.

Comment (iuliadmtru) [reply src/svymean.jl:89]: Sorry, I saw now that you just moved it a few lines below. The Stratified and SRS functions got mixed up. Try to fix that.

File: src/SurveyDesign.jl (hunk 1 above)
Comment (smishr) [reply src/SurveyDesign.jl:153]: Hi iulia in the variance formula we have an fh term, so I have kept sampfraction

File: src/ht.jl (hunk 3 above)
Comment (smishr) [reply src/ht.jl:34]: As i wrote in the issue for ht.jl, `ht_svymean` I was unable to find the correct scaling factor yet... So its not working right now

File: src/svyby.jl (hunk 4 above)
Comment (smishr) [reply src/svyby.jl:42]: Okay

File: src/svyby.jl (hunk 5 above)
Comment (smishr) [reply src/svyby.jl]: Yes that is possible

File: src/svymean.jl (hunk 9 above)
Comment (smishr) [reply src/svymean.jl:89]: Yep its still there

File: src/svymean.jl [hunk 10]
```diff
@@ -102,23 +102,39 @@ end
...
+Calculates domain mean and its std error, based on pg392 Sarndal.
 """
-function svymean(x::AbstractVector, sampfraction::AbstractVector,strata::AbstractVector,design::StratifiedSample)
-    df = DataFrame(x = x, sampfraction = sampfraction,strata = strata)
+function svymean(x::AbstractVector, popsize::AbstractVector,sampsize::AbstractVector,sampfraction::AbstractVector,strata::AbstractVector)
+    df = DataFrame(x = x, popsize = popsize, sampsize = sampsize, sampfraction = sampfraction,strata = strata)
     nsdh = []
     substrata_domain_totals = []
-    sampfractions = []
+    Nh = []
```
Comment (iuliadmtru) [inline src/svymean.jl:107]: We should try to have more readable variable names, or at least comments describing each variable. This would be the population size corresponding to each stratum, for example

File: src/svyby.jl [hunk 11]
```diff
@@ -46,5 +46,5 @@ Generate subsets of a StratifiedSample.
 function svyby(formula::Symbol, by::Symbol, design::StratifiedSample, func::Function)
     # TODO: add functionality for `formula::AbstractVector`
     gdf_domain = groupby(design.data, by)
-    return combine(gdf_domain, [formula, :sampfraction, design.strata] => ((a,b,c) -> func(a,b,c,design)) => AsTable ) 
+    return combine(gdf_domain, [formula, :popsize,:sampsize,:sampfraction, design.strata] => ((a,b,c,d,e) -> func(a,b,c,d,e)) => AsTable )
```
Comment (iuliadmtru) [inline src/svyby.jl:49]: Format the code before merging. There should be spaces after commas.

File: src/svymean.jl [hunk 12]
```diff
@@ -102,23 +102,39 @@ end
 
 """
 Inner method for `svyby` for StratifiedSample
+Calculates domain mean and its std error, based on pg392 Sarndal.
 """
-function svymean(x::AbstractVector, sampfraction::AbstractVector,strata::AbstractVector,design::StratifiedSample)
-    df = DataFrame(x = x, sampfraction = sampfraction,strata = strata)
+function svymean(x::AbstractVector, popsize::AbstractVector,sampsize::AbstractVector,sampfraction::AbstractVector,strata::AbstractVector)
```
Comment (iuliadmtru) [inline src/svymean.jl:103]: Same thing about formatting.

**Review by iuliadmtru — COMMENTED** [review]
The implementation looks good. I added some comments regarding style (code formatting and variable names/description inside comments).

**Comment by iuliadmtru** [comment]
For SRS we calculate the statistic (mean/total) and the corresponding standard error separately. We should be consistent with that, either have separate functions for `StratifiedSample` as well, or change the `SRS` implementation. I think separating them is better. In this case it would mean having a separate function for `sem` that is called by `svymean`, similar to the `SRS` functions. What do you think @smishr @ayushpatnaikgit?

**Comment by iuliadmtru** [comment]
Also, maybe we can find a better way to get the information per strata than have empty arrays that we push into. Did you try using broadcasting @smishr?

**Comment by smishr** [comment]
> For SRS we calculate the statistic (mean/total) and the corresponding standard error separately. We should be consistent with that, either have separate functions for `StratifiedSample` as well, or change the `SRS` implementation. I think separating them is better. In this case it would mean having a separate function for `sem` that is called by `svymean`, similar to the `SRS` functions. What do you think @smishr @ayushpatnaikgit?

yes please separate out now that functionality is achieved

**Comment by smishr** [comment]
> Also, maybe we can find a better way to get the information per strata than have empty arrays that we push into. Did you try using broadcasting @smishr?

For this we would have remove the for loop and add multiple combine statements to get each. At this stage, I think we should be happy for the current functionality, can look into optimising code for a later release

**Comment by iuliadmtru** [comment]
> At this stage, I think we should be happy for the current functionality, can look into optimising code for a later release

I agree.

**Comment by iuliadmtru** [comment]
> yes please separate out now that functionality is achieved

Can you please take care of that as well?

**Comment by smishr** [comment]
> > yes please separate out now that functionality is achieved
> 
> Can you please take care of that as well?

I looked through the function, subcalculations for mean and sem of mean are happening simultaneously (variance also uses `nsdh` and `substratadomaintotals`), so to do them separately will take time and care to not do `for` loops twice. @ayushpatnaikgit suggests that performance is less important for 0.2 release and we have many other engineering tasks to do. So lets leave this as is for now.

File: src/svyby.jl (hunk 4 above)
Comment (smishr) [reply src/svyby.jl:42]: I have edited the docstring in new commit

File: src/svymean.jl (hunk 6 above)
Comment (smishr) [reply src/svymean.jl:71]: removed in latest commit

File: src/svymean.jl (hunk 7 above)
Comment (smishr) [reply src/svymean.jl]: the entire fucntion has been removed
